src/reducers/form-reducer.ts

```diff
@@ -0,0 +1,59 @@
+import { FormData } from "../types"
+
+export type FormActions =
+    { type: 'set-cardholder-name', payload: { cardholderName: FormData['cardholderName'] } } |
```

Comment:
Instead using strings directly like this: 'set-cardholder-name' use enums like below because whenever you want to change such string you must change it all over the app and using enums you have to only change it here: 

```
import { FormData } from "../types";

export const enum EFormActionTypes {
  SET_CARDHOLDER_NAME = "set-cardholder-name",
  SET_CARD_NUMBER = "set-card-number",
  SET_EXP_MONTH = "set-exp-month",
  SET_EXP_YEAR = "set-exp-year",
  SET_CVC = "set-cvc",
}

export const enum EFormDataTypes {
  CARDHOLDER_NAME = "cardholderName",
  CARD_NUMBER = "cardNumber",
  EXP_MONTH = "expMonth",
  EXP_YEAR = "expYear",
  CVC = "cvc",
}

export type FormActions =
  | {
      type: EFormActionTypes.SET_CARDHOLDER_NAME;
      payload: { cardholderName: FormData[EFormDataTypes.CARDHOLDER_NAME] };
    }
  | {
      type: EFormActionTypes.SET_CARD_NUMBER;
      payload: { cardNumber: FormData[EFormDataTypes.CARD_NUMBER] };
    }
  | {
      type: EFormActionTypes.SET_EXP_MONTH;
      payload: { expMonth: FormData[EFormDataTypes.EXP_MONTH] };
    }
  | {
      type: EFormActionTypes.SET_EXP_YEAR;
      payload: { expYear: FormData[EFormDataTypes.EXP_YEAR] };
    }
  | {
      type: EFormActionTypes.SET_CVC;
      payload: { cvc: FormData[EFormDataTypes.CVC] };
    };

export const initialState: FormData = {
  cardholderName: "",
  cardNumber: "",
  expMonth: "",
  expYear: "",
  cvc: "",
};

export const formReducer = (
  state: FormData = initialState,
  action: FormActions,
) => {
  if (action.type === EFormActionTypes.SET_CARDHOLDER_NAME) {
    return {
      ...state,
      cardholderName: action.payload.cardholderName,
    };
  }

  if (action.type === EFormActionTypes.SET_CARD_NUMBER) {
    return {
      ...state,
      cardNumber: action.payload.cardNumber,
    };
  }

  if (action.type === EFormActionTypes.SET_EXP_MONTH) {
    return {
      ...state,
      expMonth: action.payload.expMonth,
    };
  }

  if (action.type === EFormActionTypes.SET_EXP_YEAR) {
    return {
      ...state,
      expYear: action.payload.expYear,
    };
  }

  if (action.type === EFormActionTypes.SET_CVC) {
    return {
      ...state,
      cvc: action.payload.cvc,
    };
  }

  return state;
};

```

Reply:
you have to update all the components that have this strings: 'set-cardholder-name' not only this but all of them, let me know if you have questions